File: ios/joinery-kit/Sources/JoineryKit/WebScreen.swift
import SwiftUI
import WebKit

/// The authenticated webview screen — how every `{type: "web"}` navigation
/// destination renders (spec § Webview contract).
///
/// - Same-origin only: off-site links (and non-web schemes) leave the app —
///   Safari, Mail, etc. Member-surface navigation stays in-webview.
/// - The native shell owns the title bar; the web page's title fills it, and
///   a toolbar chevron walks webview history when there is any.
/// - First use bridges the API session into an app-context web session; a
///   logged-out redirect (`/login`) silently re-bridges and retries once
///   before surfacing anything. A dead API key makes the re-bridge mint 401,
///   which signs the whole app out through the client handler.
/// - Pull-to-refresh, file uploads (WKWebView native pickers), downloads
///   (share-sheet hand-off), and standard loading/error/retry states.
public struct WebScreen: View {
    let title: String
    let target: String
    let client: APIClient
    let web: WebSessionCoordinator

    @State private var pageState = WebPageState()
    @State private var reloadToken = 0

    public init(title: String, target: String, client: APIClient, web: WebSessionCoordinator) {
        self.title = title
        self.target = target
        self.client = client
        self.web = web
    }

    public var body: some View {
        ZStack {
            // The webview stays inside the safe area: extending under the
            // native tab bar leaves fixed-bottom web UI (cookie consent,
            // sticky action bars) visible but untappable behind it.
            WebViewRepresentable(
                target: target,
                client: client,
                web: web,
                pageState: $pageState,
                reloadToken: reloadToken
            )

            if pageState.isLoading {
                ProgressView()
                    .accessibilityIdentifier("web_loading")
            }
            if let message = pageState.failureMessage {
                VStack(spacing: 12) {
                    Text(message)
                        .multilineTextAlignment(.center)
                        .foregroundStyle(.secondary)
                        .accessibilityIdentifier("web_error")
                    Button("Try Again") { reloadToken += 1 }
                        .buttonStyle(.borderedProminent)
                        .accessibilityIdentifier("web_retry")
                }
                .padding()
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .background(Color(uiColor: .systemBackground))
            }
        }
        .navigationTitle(pageState.pageTitle.isEmpty ? title : pageState.pageTitle)
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            if pageState.canGoBack {
                ToolbarItem(placement: .topBarLeading) {
                    Button {
                        pageState.goBackRequests += 1
                    } label: {
                        Image(systemName: "chevron.backward")
                    }
                    .accessibilityIdentifier("web_back")
                }
            }
        }
        .sheet(item: $pageState.downloadedFile) { download in
            ShareSheet(items: [download.url])
        }
    }
}

/// Webview state surfaced to the native chrome.
struct WebPageState {
    var isLoading = true
    var failureMessage: String?
    var pageTitle = ""
    var canGoBack = false
    /// Incremented by the native back chevron; the coordinator consumes it.
    var goBackRequests = 0
    var downloadedFile: DownloadedFile?
}

struct DownloadedFile: Identifiable {
    let url: URL
    var id: String { url.absoluteString }
}

private struct ShareSheet: UIViewControllerRepresentable {
    let items: [Any]
    func makeUIViewController(context: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: items, applicationActivities: nil)
    }
    func updateUIViewController(_ controller: UIActivityViewController, context: Context) {}
}

// MARK: - WKWebView wrapper

private struct WebViewRepresentable: UIViewRepresentable {
    let target: String
    let client: APIClient
    let web: WebSessionCoordinator
    @Binding var pageState: WebPageState
    let reloadToken: Int

    func makeCoordinator() -> WebCoordinator {
        WebCoordinator(target: target, client: client, web: web)
    }

    func makeUIView(context: Context) -> WKWebView {
        let configuration = WKWebViewConfiguration()
        // The shared default store carries the bridged session cookie across
        // every webview and across launches.
        configuration.websiteDataStore = .default()
        configuration.allowsInlineMediaPlayback = true

        let webView = WKWebView(frame: .zero, configuration: configuration)
        webView.navigationDelegate = context.coordinator
        webView.uiDelegate = context.coordinator
        webView.allowsBackForwardNavigationGestures = true
        webView.scrollView.refreshControl = context.coordinator.makeRefreshControl()

        context.coordinator.attach(webView: webView) { update in
            pageState = update(pageState)
        }
        context.coordinator.startInitialLoad()
        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        context.coordinator.onStateWrite = { update in
            pageState = update(pageState)
        }
        context.coordinator.consumeGoBack(requests: pageState.goBackRequests)
        context.coordinator.retryIfRequested(token: reloadToken)
    }
}

/// All webview policy lives here: bridging, logged-out detection, the link
/// policy, downloads, and refresh.
@MainActor
final class WebCoordinator: NSObject {
    private let target: String
    private let client: APIClient
    private let web: WebSessionCoordinator

    private weak var webView: WKWebView?
    var onStateWrite: ((@escaping (WebPageState) -> WebPageState) -> Void)?

    /// The path the user is meant to be on — re-bridges land back here.
    private var intendedPath: String
    /// One silent re-bridge per load attempt; reset on success and refresh.
    private var didRebridge = false
    /// Set while the current navigation is a bridge URL, so its completion
    /// marks the shared coordinator bridged.
    private var bridging = false

    private var titleObservation: NSKeyValueObservation?
    private var backObservation: NSKeyValueObservation?
    private var consumedGoBacks = 0
    private var lastReloadToken = 0
    private var downloadDestination: URL?

    init(target: String, client: APIClient, web: WebSessionCoordinator) {
        self.target = target
        self.client = client
        self.web = web
        self.intendedPath = target
        super.init()
    }

    func attach(webView: WKWebView, stateWriter: @escaping ((@escaping (WebPageState) -> WebPageState) -> Void)) {
        self.webView = webView
        self.onStateWrite = stateWriter
        titleObservation = webView.observe(\.title, options: [.new]) { [weak self] view, _ in
            let title = view.title ?? ""
            Task { @MainActor in self?.write { state in var s = state; s.pageTitle = title; return s } }
        }
        backObservation = webView.observe(\.canGoBack, options: [.new]) { [weak self] view, _ in
            let canGoBack = view.canGoBack
            Task { @MainActor in self?.write { state in var s = state; s.canGoBack = canGoBack; return s } }
        }
    }

    private func write(_ update: @escaping (WebPageState) -> WebPageState) {
        onStateWrite?(update)
    }

    func makeRefreshControl() -> UIRefreshControl {
        let control = UIRefreshControl()
        control.addTarget(self, action: #selector(pullToRefresh), for: .valueChanged)
        return control
    }

    @objc private func pullToRefresh() {
        didRebridge = false
        webView?.reload()
    }

    func consumeGoBack(requests: Int) {
        guard requests > consumedGoBacks else { return }
        consumedGoBacks = requests
        webView?.goBack()
    }

    func retryIfRequested(token: Int) {
        guard token != lastReloadToken else { return }
        lastReloadToken = token
        didRebridge = false
        write { state in var s = state; s.failureMessage = nil; s.isLoading = true; return s }
        startInitialLoad()
    }

    // MARK: Loading

    func startInitialLoad() {
        Task { await load(path: intendedPath) }
    }

    private func load(path: String) async {
        guard let webView else { return }
        if web.bridged, let direct = web.pageURL(for: path) {
            webView.load(URLRequest(url: direct))
            return
        }
        await bridge(to: path)
    }

    /// Mint a fresh single-use bridge token and drive the webview through it.
    /// A 401 here means the API key is dead — the client handler signs the
    /// whole app out, so this screen only needs to stop quietly.
    private func bridge(to path: String) async {
        guard let webView else { return }
        do {
            let bridgeURL = try await web.bridgeURL(for: path)
            bridging = true
            webView.load(URLRequest(url: bridgeURL))
        } catch let error as JoineryAPIError {
            if case .authentication = error { return }
            fail(error.displayMessage)
        } catch {
            fail("This page could not be loaded.")
        }
    }

    private func fail(_ message: String) {
        write { state in
            var s = state
            s.isLoading = false
            s.failureMessage = message
            return s
        }
        webView?.scrollView.refreshControl?.endRefreshing()
    }

    /// The logged-out signal: the site 302s member pages to `/login` when the
    /// bridged session is gone. Silently re-bridge back to where the user was
    /// headed; only a second failure in the same attempt surfaces.
    private func handleLoggedOut() {
        if didRebridge {
            fail("Your session has expired. Pull to refresh to try again.")
            return
        }
        didRebridge = true
        let path = intendedPath
        Task { await bridge(to: path) }
    }

    private func isSameOrigin(_ url: URL) -> Bool {
        guard let host = url.host, let baseHost = web.baseHost else { return false }
        return host == baseHost
    }
}

// MARK: WKNavigationDelegate

extension WebCoordinator: WKNavigationDelegate {

    func webView(_ webView: WKWebView,
                 decidePolicyFor navigationAction: WKNavigationAction,
                 decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        guard let url = navigationAction.request.url else {
            decisionHandler(.cancel)
            return
        }

        // Subframes (embedded payment fields, media) are the page's business.
        guard navigationAction.targetFrame?.isMainFrame != false else {
            decisionHandler(.allow)
            return
        }

        // Non-web schemes (mailto:, tel:, app links) leave the app.
        if let scheme = url.scheme?.lowercased(), scheme != "http", scheme != "https" {
            UIApplication.shared.open(url)
            decisionHandler(.cancel)
            return
        }

        // Off-site: Safari (spec — same-origin only in the webview).
        if !isSameOrigin(url) {
            UIApplication.shared.open(url)
            decisionHandler(.cancel)
            return
        }

        // Logged-out redirect → silent re-bridge.
        if url.path == "/login" || url.path.hasPrefix("/login/") {
            decisionHandler(.cancel)
            handleLoggedOut()
            return
        }

        // Track where the user is headed so a mid-session re-bridge lands
        // back on the right page (bridge plumbing itself doesn't count).
        if url.path != "/app_bridge" {
            var path = url.path.isEmpty ? "/" : url.path
            if let query = url.query, !query.isEmpty { path += "?" + query }
            intendedPath = path
        }
        decisionHandler(.allow)
    }

    func webView(_ webView: WKWebView,
                 decidePolicyFor navigationResponse: WKNavigationResponse,
                 decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void) {
        if let http = navigationResponse.response as? HTTPURLResponse {
            // An expired bridge token renders 410 — mint a fresh one.
            if http.statusCode == 410, http.url?.path == "/app_bridge" {
                decisionHandler(.cancel)
                handleLoggedOut()
                return
            }
        }
        // Content the webview can't render (files) becomes a download.
        if !navigationResponse.canShowMIMEType {
            decisionHandler(.download)
            return
        }
        decisionHandler(.allow)
    }

    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        write { state in var s = state; s.isLoading = true; s.failureMessage = nil; return s }
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        if bridging {
            bridging = false
            web.markBridged()
        }
        didRebridge = false
        write { state in var s = state; s.isLoading = false; return s }
        webView.scrollView.refreshControl?.endRefreshing()
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        finishWithError(error)
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        finishWithError(error)
    }

    private func finishWithError(_ error: Error) {
        let nsError = error as NSError
        // Cancellations (our own policy cancels, mid-flight reloads) are not
        // failures the user should see.
        if nsError.domain == NSURLErrorDomain && nsError.code == NSURLErrorCancelled { return }
        if nsError.domain == "WebKitErrorDomain" && nsError.code == 102 { return } // frame load interrupted by policy
        fail(nsError.code == NSURLErrorNotConnectedToInternet
             ? "You appear to be offline. Pull to refresh when you're back online."
             : "This page could not be loaded.")
    }

    func webView(_ webView: WKWebView,
                 navigationResponse: WKNavigationResponse,
                 didBecome download: WKDownload) {
        download.delegate = self
    }
}

// MARK: WKUIDelegate (target=_blank, JS dialogs)

extension WebCoordinator: WKUIDelegate {
    func webView(_ webView: WKWebView,
                 createWebViewWith configuration: WKWebViewConfiguration,
                 for navigationAction: WKNavigationAction,
                 windowFeatures: WKWindowFeatures) -> WKWebView? {
        if let url = navigationAction.request.url {
            if isSameOrigin(url) {
                webView.load(URLRequest(url: url))
            } else {
                UIApplication.shared.open(url)
            }
        }
        return nil
    }

    // The member surface uses alert()/confirm() (e.g. "Revoke All" asks for
    // confirmation). WKWebView drops them silently unless the UI delegate
    // presents native panels.

    func webView(_ webView: WKWebView,
                 runJavaScriptAlertPanelWithMessage message: String,
                 initiatedByFrame frame: WKFrameInfo,
                 completionHandler: @escaping () -> Void) {
        let alert = UIAlertController(title: nil, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default) { _ in completionHandler() })
        guard present(alert) else { completionHandler(); return }
    }

    func webView(_ webView: WKWebView,
                 runJavaScriptConfirmPanelWithMessage message: String,
                 initiatedByFrame frame: WKFrameInfo,
                 completionHandler: @escaping (Bool) -> Void) {
        let alert = UIAlertController(title: nil, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel) { _ in completionHandler(false) })
        alert.addAction(UIAlertAction(title: "OK", style: .default) { _ in completionHandler(true) })
        guard present(alert) else { completionHandler(false); return }
    }

    func webView(_ webView: WKWebView,
                 runJavaScriptTextInputPanelWithPrompt prompt: String,
                 defaultText: String?,
                 initiatedByFrame frame: WKFrameInfo,
                 completionHandler: @escaping (String?) -> Void) {
        let alert = UIAlertController(title: nil, message: prompt, preferredStyle: .alert)
        alert.addTextField { $0.text = defaultText }
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel) { _ in completionHandler(nil) })
        alert.addAction(UIAlertAction(title: "OK", style: .default) { [weak alert] _ in
            completionHandler(alert?.textFields?.first?.text ?? "")
        })
        guard present(alert) else { completionHandler(nil); return }
    }

    /// Present on the frontmost view controller; false when none is found.
    private func present(_ alert: UIAlertController) -> Bool {
        let scene = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
        guard var top = scene?.keyWindow?.rootViewController else { return false }
        while let presented = top.presentedViewController { top = presented }
        top.present(alert, animated: true)
        return true
    }
}

// MARK: WKDownloadDelegate (share-sheet hand-off)

extension WebCoordinator: WKDownloadDelegate {
    func download(_ download: WKDownload,
                  decideDestinationUsing response: URLResponse,
                  suggestedFilename: String,
                  completionHandler: @escaping (URL?) -> Void) {
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent("JoineryDownloads-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let destination = directory.appendingPathComponent(suggestedFilename)
        downloadDestination = destination
        completionHandler(destination)
    }

    func downloadDidFinish(_ download: WKDownload) {
        guard let fileURL = downloadDestination else { return }
        write { state in
            var s = state
            s.isLoading = false
            s.downloadedFile = DownloadedFile(url: fileURL)
            return s
        }
        webView?.scrollView.refreshControl?.endRefreshing()
    }

    func download(_ download: WKDownload, didFailWithError error: Error, resumeData: Data?) {
        fail("The download failed.")
    }
}

File: android/joinery-android/src/main/java/com/getjoinery/android/WebScreen.kt
package com.getjoinery.android

import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.view.ViewGroup
import android.webkit.CookieManager
import android.webkit.URLUtil
import android.webkit.ValueCallback
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.compose.BackHandler
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.browser.customtabs.CustomTabsIntent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/**
 * Webview state surfaced to the native chrome. The shell reads [pageTitle] and
 * [canGoBack] for the app bar; [WebScreen] writes all of it.
 */
class WebPageState {
    var isLoading by mutableStateOf(true)
    var failureMessage by mutableStateOf<String?>(null)
    var pageTitle by mutableStateOf("")
    var canGoBack by mutableStateOf(false)
}

/**
 * The authenticated webview — how every `{type: "web"}` navigation destination
 * renders (spec § Webview contract).
 *
 * - Same-origin only: off-site links (Custom Tabs) and non-web schemes (mailto,
 *   tel) leave the app; member-surface navigation stays in-webview.
 * - First use bridges the API session into an app-context web session; a
 *   logged-out redirect (`/login`) or an expired bridge token (410) silently
 *   re-bridges and retries once before surfacing anything. A dead API key makes
 *   the re-bridge mint 401, which signs the whole app out through the client
 *   handler.
 * - System back walks webview history first (this screen's [BackHandler]), then
 *   the shell's navigation. Pull-to-refresh, file uploads (SAF picker),
 *   downloads (DownloadManager), and loading/error/offline-retry states.
 *
 * The caller owns the app bar and reads the page title / back affordance from
 * [state].
 */
@Composable
fun WebScreen(
    target: String,
    client: ApiClient,
    web: WebSessionCoordinator,
    state: WebPageState,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val controller = remember(target) { WebController(context, target, client, web, state, scope) }

    val fileChooser = rememberLauncherForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        controller.onFileChooserResult(result.resultCode, result.data)
    }

    // System back walks webview history while there is any; when there is none,
    // this handler is disabled and the shell's navigation-back takes over.
    BackHandler(enabled = state.canGoBack) { controller.goBack() }

    Box(modifier.fillMaxSize()) {
        AndroidView(
            // The persistent marker that web content is hosted here — the
            // web_loading spinner below is transient, so tests key on this.
            modifier = Modifier.fillMaxSize().testTag("web_view"),
            factory = { ctx ->
                val swipe = SwipeRefreshLayout(ctx)
                val webView = WebView(ctx).apply {
                    layoutParams = ViewGroup.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT,
                    )
                    settings.javaScriptEnabled = true
                    settings.domStorageEnabled = true
                    settings.useWideViewPort = true
                    settings.loadWithOverviewMode = true
                    settings.mediaPlaybackRequiresUserGesture = false
                    webViewClient = controller.webViewClient()
                    webChromeClient = controller.webChromeClient(fileChooser)
                    setDownloadListener(controller.downloadListener())
                }
                CookieManager.getInstance().apply {
                    setAcceptCookie(true)
                    setAcceptThirdPartyCookies(webView, true)
                }
                swipe.addView(webView)
                swipe.setOnRefreshListener { controller.refresh() }
                controller.attach(webView, swipe)
                controller.startInitialLoad()
                swipe
            },
        )

        if (state.isLoading) {
            CircularProgressIndicator(
                Modifier.align(Alignment.Center).testTag("web_loading")
            )
        }
        state.failureMessage?.let { message ->
            Column(
                Modifier.align(Alignment.Center).padding(24.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Text(
                    message,
                    textAlign = TextAlign.Center,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.testTag("web_error"),
                )
                Button(onClick = { controller.retry() }, modifier = Modifier.testTag("web_retry")) {
                    Text("Try Again")
                }
            }
        }
    }
}

/**
 * All webview policy lives here: bridging, logged-out detection, the link
 * policy, downloads, file uploads, and refresh. Not a composable — it holds the
 * live [WebView] and drives it imperatively from the delegate callbacks.
 */
private class WebController(
    private val context: Context,
    target: String,
    private val client: ApiClient,
    private val web: WebSessionCoordinator,
    private val state: WebPageState,
    private val scope: CoroutineScope,
) {
    private var webView: WebView? = null
    private var swipe: SwipeRefreshLayout? = null

    /** The path the user is meant to be on — re-bridges land back here. */
    private var intendedPath: String = target

    /** One silent re-bridge per load attempt; reset on success and refresh. */
    private var didRebridge = false

    /** Set while the current navigation is a bridge URL, so its completion marks
     *  the shared coordinator bridged. */
    private var bridging = false

    private var filePathCallback: ValueCallback<Array<Uri>>? = null

    fun attach(webView: WebView, swipe: SwipeRefreshLayout) {
        this.webView = webView
        this.swipe = swipe
    }

    // MARK: Loading

    fun startInitialLoad() {
        scope.launch { load(intendedPath) }
    }

    private suspend fun load(path: String) {
        val view = webView ?: return
        if (web.bridged) {
            view.loadUrl(web.pageUrl(path))
            return
        }
        bridge(path)
    }

    /** Mint a fresh single-use bridge token and drive the webview through it. A
     *  401 here means the API key is dead — the client handler signs the whole
     *  app out, so this screen only needs to stop quietly. */
    private suspend fun bridge(path: String) {
        val view = webView ?: return
        try {
            val bridgeUrl = web.bridgeUrl(path)
            bridging = true
            view.loadUrl(bridgeUrl)
        } catch (e: JoineryApiError) {
            if (e is JoineryApiError.Authentication) return
            fail(e.displayMessage)
        } catch (e: Exception) {
            fail("This page could not be loaded.")
        }
    }

    /** The logged-out signal: the site 302s member pages to `/login` when the
     *  bridged session is gone, and an expired bridge token renders 410. Silently
     *  re-bridge back to where the user was headed; only a second failure in the
     *  same attempt surfaces. */
    private fun handleLoggedOut() {
        if (didRebridge) {
            fail("Your session has expired. Pull down to refresh and try again.")
            return
        }
        didRebridge = true
        val path = intendedPath
        scope.launch { bridge(path) }
    }

    private fun fail(message: String) {
        state.isLoading = false
        state.failureMessage = message
        swipe?.isRefreshing = false
    }

    fun retry() {
        didRebridge = false
        state.failureMessage = null
        state.isLoading = true
        startInitialLoad()
    }

    fun refresh() {
        didRebridge = false
        if (web.bridged) webView?.reload() else startInitialLoad()
    }

    fun goBack() {
        webView?.goBack()
    }

    private fun isSameOrigin(uri: Uri): Boolean {
        val host = uri.host ?: return false
        val base = web.baseHost ?: return false
        return host == base
    }

    private fun openExternal(uri: Uri) {
        runCatching { context.startActivity(Intent(Intent.ACTION_VIEW, uri)) }
    }

    private fun openExternalTab(uri: Uri) {
        runCatching { CustomTabsIntent.Builder().build().launchUrl(context, uri) }
            .onFailure { openExternal(uri) }
    }

    // MARK: Delegates

    fun webViewClient(): WebViewClient = object : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            val url = request.url
            // Subframes (embedded payment fields, media) are the page's business.
            if (!request.isForMainFrame) return false

            val scheme = url.scheme?.lowercase()
            if (scheme != "http" && scheme != "https") {
                openExternal(url)
                return true
            }
            // Off-site: Custom Tabs (spec — same-origin only in the webview).
            if (!isSameOrigin(url)) {
                openExternalTab(url)
                return true
            }
            val path = url.path ?: "/"
            // Logged-out redirect → silent re-bridge.
            if (path == "/login" || path.startsWith("/login/")) {
                handleLoggedOut()
                return true
            }
            // Track where the user is headed so a mid-session re-bridge lands back
            // on the right page (bridge plumbing itself doesn't count).
            if (path != "/app_bridge") {
                val query = url.query
                intendedPath = if (query.isNullOrEmpty()) path else "$path?$query"
            }
            return false
        }

        override fun onPageStarted(view: WebView?, url: String?, favicon: android.graphics.Bitmap?) {
            state.isLoading = true
            state.failureMessage = null
        }

        override fun onPageFinished(view: WebView?, url: String?) {
            if (bridging) {
                bridging = false
                web.markBridged()
            }
            didRebridge = false
            state.isLoading = false
            state.canGoBack = view?.canGoBack() ?: false
            swipe?.isRefreshing = false
        }

        override fun doUpdateVisitedHistory(view: WebView?, url: String?, isReload: Boolean) {
            state.canGoBack = view?.canGoBack() ?: false
        }

        override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
            if (request?.isForMainFrame != true) return
            val offline = when (error?.errorCode) {
                WebViewClient.ERROR_HOST_LOOKUP,
                WebViewClient.ERROR_CONNECT,
                WebViewClient.ERROR_TIMEOUT -> true
                else -> false
            }
            fail(
                if (offline) "You appear to be offline. Pull down to refresh when you're back online."
                else "This page could not be loaded."
            )
        }

        override fun onReceivedHttpError(
            view: WebView?,
            request: WebResourceRequest?,
            errorResponse: WebResourceResponse?,
        ) {
            // An expired bridge token renders 410 — mint a fresh one.
            if (request?.isForMainFrame == true &&
                request.url.path == "/app_bridge" &&
                errorResponse?.statusCode == 410
            ) {
                handleLoggedOut()
            }
        }
    }

    fun webChromeClient(launcher: ActivityResultLauncher<Intent>): WebChromeClient =
        object : WebChromeClient() {
            override fun onReceivedTitle(view: WebView?, title: String?) {
                state.pageTitle = title ?: ""
            }

            override fun onShowFileChooser(
                webView: WebView?,
                callback: ValueCallback<Array<Uri>>?,
                params: FileChooserParams?,
            ): Boolean {
                filePathCallback?.onReceiveValue(null)
                filePathCallback = callback
                return try {
                    val intent = params?.createIntent()
                    if (intent != null) {
                        launcher.launch(intent)
                        true
                    } else {
                        filePathCallback = null
                        false
                    }
                } catch (e: Exception) {
                    filePathCallback = null
                    false
                }
            }
        }

    fun onFileChooserResult(resultCode: Int, data: Intent?) {
        val callback = filePathCallback ?: return
        filePathCallback = null
        callback.onReceiveValue(WebChromeClient.FileChooserParams.parseResult(resultCode, data))
    }

    /** Content the webview can't render (PDFs, attachments) hands off to
     *  DownloadManager, carrying the bridged session cookie so authenticated
     *  files download. Lands in the app's external Downloads dir (no runtime
     *  permission on any supported API) with a completion notification. */
    fun downloadListener() = android.webkit.DownloadListener { url, userAgent, contentDisposition, mimeType, _ ->
        runCatching {
            val request = DownloadManager.Request(Uri.parse(url))
            CookieManager.getInstance().getCookie(url)?.let { request.addRequestHeader("Cookie", it) }
            request.addRequestHeader("User-Agent", userAgent)
            val name = URLUtil.guessFileName(url, contentDisposition, mimeType)
            request.setMimeType(mimeType)
            request.setTitle(name)
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
            request.setDestinationInExternalFilesDir(context, Environment.DIRECTORY_DOWNLOADS, name)
            (context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager).enqueue(request)
        }
    }
}
